File: buy.py
import requests
import time
import json
from typing import Optional, Dict, Any

# Configuration
API_BASE_URL = "https://api.vortena.xyz"
PRIVATE_KEY = "YOUR_PRIVATE_KEY_HERE"  # Replace with your Base58 encoded private key
TOKEN_MINT = "TOKEN_MINT_ADDRESS_HERE"  # Replace with token contract address

# Trading parameters
TRADE_CONFIG = {
    "amount": 0.01,        # SOL amount to spend
    "slippage": 5,         # 5% slippage tolerance
    "priority_fee": 0.003  # Priority fee in SOL
}

class VortenaTrader:
    """VORTENA API Trading Client"""
    
    def __init__(self, private_key: str, base_url: str = API_BASE_URL):
        self.private_key = private_key
        self.base_url = base_url
        self.session = requests.Session()
        self.session.timeout = 30
    
    def buy_tokens(self, mint: str, amount: float, slippage: int = 5, priority_fee: float = 0.003) -> Optional[Dict[str, Any]]:
        """
        Execute a buy trade
        
        Args:
            mint: Token mint address
            amount: SOL amount to spend
            slippage: Slippage tolerance in percentage
            priority_fee: Priority fee in SOL
            
        Returns:
            Trade result dictionary or None if failed
        """
        url = f"{self.base_url}/trade/buy/{mint}/{amount}/{slippage}/{priority_fee}/{self.private_key}"
        
        try:
            print("🔄 Executing buy trade...")
            print(f"Token: {mint}")
            print(f"Amount: {amount} SOL")
            print(f"Slippage: {slippage}%")
            print(f"Priority Fee: {priority_fee} SOL")
            
            response = self.session.get(url)
            response.raise_for_status()
            result = response.json()
            
            if result.get("success"):
                print("✅ Trade successful!")
                print(f"Transaction Hash: {result.get('tx_hash', 'N/A')}")
                print(f"Solscan URL: {result.get('solscan_url', 'N/A')}")
                print(f"Amount Received: {result.get('amount_received', 'N/A')} tokens")
                return result
            else:
                print(f"❌ Trade failed: {result.get('error', 'Unknown error')}")
                return None
                
        except requests.exceptions.RequestException as e:
            print(f"❌ Request error: {str(e)}")
            return None
        except json.JSONDecodeError:
            print("❌ Invalid JSON response")
            return None
        except Exception as e:
            print(f"❌ Unexpected error: {str(e)}")
            return None
    
    def buy_with_retry(self, mint: str, amount: float, max_retries: int = 3) -> Optional[Dict[str, Any]]:
        """
        Buy tokens with retry logic
        
        Args:
            mint: Token mint address
            amount: SOL amount to spend
            max_retries: Maximum number of retry attempts
            
        Returns:
            Trade result dictionary or None if all attempts failed
        """
        for attempt in range(1, max_retries + 1):
            print(f"\n🔄 Attempt {attempt}/{max_retries}")
            
            try:
                # Increase slippage on retries
                slippage = TRADE_CONFIG["slippage"] + (attempt - 1) * 2
                result = self.buy_tokens(mint, amount, slippage, TRADE_CONFIG["priority_fee"])
                
                if result and result.get("success"):
                    return result
                
                if attempt < max_retries:
                    print("⏳ Waiting 3 seconds before retry...")
                    time.sleep(3)
                    
            except Exception as e:
                print(f"❌ Attempt {attempt} failed: {str(e)}")
                
                if attempt < max_retries:
                    print("⏳ Waiting 3 seconds before retry...")
                    time.sleep(3)
        
        print("❌ All retry attempts failed")
        return None

def validate_params(mint: str, amount: float, slippage: int, priority_fee: float):
    """Validate trading parameters"""
    if not mint or mint == "TOKEN_MINT_ADDRESS_HERE":
        raise ValueError("Please provide a valid token mint address")
    
    if not PRIVATE_KEY or PRIVATE_KEY == "YOUR_PRIVATE_KEY_HERE":
        raise ValueError("Please provide your private key")
    
    if amount <= 0:
        raise ValueError("Amount must be greater than 0")
    
    if slippage < 0.1 or slippage > 50:
        raise ValueError("Slippage must be between 0.1% and 50%")
    
    if priority_fee < 0:
        raise ValueError("Priority fee must be non-negative")

def main():
    """Main execution function"""
    try:
        # Validate parameters
        validate_params(
            TOKEN_MINT, 
            TRADE_CONFIG["amount"], 
            TRADE_CONFIG["slippage"], 
            TRADE_CONFIG["priority_fee"]
        )
        
        print("🚀 Starting VORTENA Buy Trade Example")
        print("=====================================")
        
        # Initialize trader
        trader = VortenaTrader(PRIVATE_KEY)
        
        # Execute trade with retry logic
        result = trader.buy_with_retry(TOKEN_MINT, TRADE_CONFIG["amount"])
        
        if result:
            print("\n🎉 Trade completed successfully!")
            print(f"Final result: {json.dumps(result, indent=2)}")
        else:
            print("\n💥 Trade failed after all retries")
            
    except ValueError as e:
        print(f"❌ Configuration error: {str(e)}")
        print("\n📝 Please check your configuration:")
        print("1. Set your private key in PRIVATE_KEY variable")
        print("2. Set the token mint address in TOKEN_MINT variable")
        print("3. Adjust trading parameters in TRADE_CONFIG if needed")
    except Exception as e:
        print(f"❌ Unexpected error: {str(e)}")

if __name__ == "__main__":
    main() 

File: buy.js
const https = require('https');

// Configuration
const API_BASE_URL = 'https://api.vortena.xyz';
const PRIVATE_KEY = 'YOUR_PRIVATE_KEY_HERE'; // Replace with your Base58 encoded private key
const TOKEN_MINT = 'TOKEN_MINT_ADDRESS_HERE'; // Replace with token contract address

// Trading parameters
const TRADE_CONFIG = {
  amount: 0.01,        // SOL amount to spend
  slippage: 5,         // 5% slippage tolerance
  priorityFee: 0.003   // Priority fee in SOL
};

/**
 * Execute a buy trade
 */
async function buyTokens(mint, amount, slippage = 5, priorityFee = 0.003) {
  const url = `${API_BASE_URL}/trade/buy/${mint}/${amount}/${slippage}/${priorityFee}/${PRIVATE_KEY}`;
  
  try {
    console.log('🔄 Executing buy trade...');
    console.log(`Token: ${mint}`);
    console.log(`Amount: ${amount} SOL`);
    console.log(`Slippage: ${slippage}%`);
    console.log(`Priority Fee: ${priorityFee} SOL`);
    
    const response = await fetch(url);
    const result = await response.json();
    
    if (result.success) {
      console.log('✅ Trade successful!');
      console.log(`Transaction Hash: ${result.tx_hash || 'N/A'}`);
      console.log(`Solscan URL: ${result.solscan_url || 'N/A'}`);
      console.log(`Amount Received: ${result.amount_received || 'N/A'} tokens`);
      return result;
    } else {
      console.error('❌ Trade failed:', result.error);
      return null;
    }
  } catch (error) {
    console.error('❌ Error executing trade:', error.message);
    return null;
  }
}

/**
 * Buy tokens with retry logic
 */
async function buyWithRetry(mint, amount, maxRetries = 3) {
  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    console.log(`\n🔄 Attempt ${attempt}/${maxRetries}`);
    
    try {
      // Increase slippage on retries
      const slippage = TRADE_CONFIG.slippage + (attempt - 1) * 2;
      const result = await buyTokens(mint, amount, slippage, TRADE_CONFIG.priorityFee);
      
      if (result && result.success) {
        return result;
      }
      
      if (attempt < maxRetries) {
        console.log(`⏳ Waiting 3 seconds before retry...`);
        await new Promise(resolve => setTimeout(resolve, 3000));
      }
    } catch (error) {
      console.error(`❌ Attempt ${attempt} failed:`, error.message);
      
      if (attempt < maxRetries) {
        console.log(`⏳ Waiting 3 seconds before retry...`);
        await new Promise(resolve => setTimeout(resolve, 3000));
      }
    }
  }
  
  console.error('❌ All retry attempts failed');
  return null;
}

/**
 * Validate trade parameters
 */
function validateParams(mint, amount, slippage, priorityFee) {
  if (!mint || mint === 'TOKEN_MINT_ADDRESS_HERE') {
    throw new Error('Please provide a valid token mint address');
  }
  
  if (!PRIVATE_KEY || PRIVATE_KEY === 'YOUR_PRIVATE_KEY_HERE') {
    throw new Error('Please provide your private key');
  }
  
  if (amount <= 0) {
    throw new Error('Amount must be greater than 0');
  }
  
  if (slippage < 0.1 || slippage > 50) {
    throw new Error('Slippage must be between 0.1% and 50%');
  }
  
  if (priorityFee < 0) {
    throw new Error('Priority fee must be non-negative');
  }
}

/**
 * Main execution function
 */
async function main() {
  try {
    // Validate parameters
    validateParams(TOKEN_MINT, TRADE_CONFIG.amount, TRADE_CONFIG.slippage, TRADE_CONFIG.priorityFee);
    
    console.log('🚀 Starting VORTENA Buy Trade Example');
    console.log('=====================================');
    
    // Execute trade with retry logic
    const result = await buyWithRetry(TOKEN_MINT, TRADE_CONFIG.amount);
    
    if (result) {
      console.log('\n🎉 Trade completed successfully!');
    } else {
      console.log('\n💥 Trade failed after all retries');
    }
    
  } catch (error) {
    console.error('❌ Configuration error:', error.message);
    console.log('\n📝 Please check your configuration:');
    console.log('1. Set your private key in PRIVATE_KEY variable');
    console.log('2. Set the token mint address in TOKEN_MINT variable');
    console.log('3. Adjust trading parameters in TRADE_CONFIG if needed');
  }
}

// Run the example
if (require.main === module) {
  main().catch(console.error);
}

module.exports = { buyTokens, buyWithRetry }; 